### DISCUSSION TITLE
Go 语言的栈内存和逃逸分析 | Go 语言设计与实现 · /golang-stack-management

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#601] [Announcements]
https://draveness.me/golang-stack-management/ 

 Go 语言的栈内存和逃逸分析

**Comment by Alien-Leon** [comment]
7.3.1设计原理中的逃逸分析一节中
> 需要分配到栈上的对象分配到了栈上 — 悬挂指针、影响内存安全；

应该是

> 不需要分配到栈上的对象分配到了栈上 — 悬挂指针、影响内存安全；

**Comment by draveness** [comment]
> 7.3.1设计原理中的逃逸分析一节中
> 
> > 需要分配到栈上的对象分配到了栈上 — 悬挂指针、影响内存安全；
> 
> 应该是
> 
> > 不需要分配到栈上的对象分配到了栈上 — 悬挂指针、影响内存安全；

需要分配到**堆**上的分配到了栈上，栈被释放后会影响安全

**Comment by starsz** [comment]
```
int *dangling_pointer() {
    int i = 2;
    return &2;
}
```

这块代码好像错了，应该是

```
int *dangling_pointer() {
    int i = 2;
    return &i;
}
```

**Comment by draveness** [comment]
> ```
> int *dangling_pointer() {
>     int i = 2;
>     return &2;
> }
> ```
> 
> 这块代码好像错了，应该是
> 
> ```
> int *dangling_pointer() {
>     int i = 2;
>     return &i;
> }
> ```

修复了，感谢

**Comment by fengwutianya** [comment]
adjustpointer时怎么知道 这是栈上内容是指针而不是普通数据？

**Comment by draveness** [comment]
> adjustpointer时怎么知道 这是栈上内容是指针而不是普通数据？

FYI

> Each word has 2 bits associated with it to tell you if it was a scalar or a pointer inside that word
>
> --- https://blog.golang.org/ismmkeynote

**Comment by fengwutianya** [comment]
> > adjustpointer时怎么知道 这是栈上内容是指针而不是普通数据？
> 
> FYI
> 
> > Each word has 2 bits associated with it to tell you if it was a scalar or a pointer inside that word
> > --- https://blog.golang.org/ismmkeynote

but there is no headers for the object, where did they put the object's metadata and what exactly is that "each word"

**Comment by fengwutianya** [comment]
> > adjustpointer时怎么知道 这是栈上内容是指针而不是普通数据？
> 
> FYI
> 
> > Each word has 2 bits associated with it to tell you if it was a scalar or a pointer inside that word
> > --- https://blog.golang.org/ismmkeynote

I found out that there is a metadata for every function which is stored in funcinfo

**Comment by discipline-yyl** [comment]
“”如果要触发栈的缩容，新栈的大小会使原始栈的一半“”
应该是 “会是”，哈哈，太专业的错误识别不了，只能帮您找错别字了

**Comment by draveness** [comment]
> “”如果要触发栈的缩容，新栈的大小会使原始栈的一半“”
> 应该是 “会是”，哈哈，太专业的错误识别不了，只能帮您找错别字了

*2020-07-10 UPDATES: 已修复*

**Comment by xiao7737** [comment]
>如果要触发栈的缩容，新栈的大小会是原始栈的一半，不过如果新栈的大小低于程序的最地限制 2KB，那么缩容的过程就会停止。

最低限制2KB，最地-->最低

---

*2020-07-18 UPDATES：已修复*

**Comment by yanjinbin** [comment]
这篇 我读下来,还是没有理解清楚分段栈和连续栈的区别 , 疑问🤔️都不是要扩容和缩容么? 
至于分段栈的问题 ,   和     连续栈  “ 因为 采用 分配 更大的栈内存空间,并且复制迁移到新空间销毁旧空间,避免上述问题” 
我没读出两者之间的 本质区别或者联系

**Reply by ChangwangChen** [reply]
看起来就是链表和数组的区别

**Comment by yanjinbin** [comment]
> 分段栈机制虽然能够按需为当前 Goroutine 分配内存并且及时减少内存的占用，但是它也存在两个比较大的问题：

> 如果当前 Goroutine 的栈几乎充满，那么任意的函数调用都会触发栈的扩容，当函数返回后又会触发栈的收缩，如果在一个循环中调用函数，栈的分配和释放就会造成巨大的额外开销，这被称为热分裂问题（Hot split）；
> 一旦 Goroutine 使用的内存越过了分段栈的扩缩容阈值，运行时就会触发栈的扩容和缩容，带来额外的工作量；

这里讲的比较模糊,没看懂  1 巨大的额外开销是什么?   2 额外的工作量是什么,这个在连续栈上也要缩扩容,有区别么?
所以提出之前那个疑问

**Comment by draveness** [comment]
> > 分段栈机制虽然能够按需为当前 Goroutine 分配内存并且及时减少内存的占用，但是它也存在两个比较大的问题：
> 
> > 如果当前 Goroutine 的栈几乎充满，那么任意的函数调用都会触发栈的扩容，当函数返回后又会触发栈的收缩，如果在一个循环中调用函数，栈的分配和释放就会造成巨大的额外开销，这被称为热分裂问题（Hot split）；
> > 一旦 Goroutine 使用的内存越过了分段栈的扩缩容阈值，运行时就会触发栈的扩容和缩容，带来额外的工作量；
> 
> 这里讲的比较模糊,没看懂 1 巨大的额外开销是什么? 

栈的分配和释放

> 2 额外的工作量是什么,这个在连续栈上也要缩扩容,有区别么?
> 所以提出之前那个疑问

分段栈中包含扩缩容的阈值，所以栈的大小可能一直都在变

可以看下连续栈的 [设计文档](https://docs.google.com/document/d/1wAaf1rYoM4S4gtnPh0zOlGzWtrZFQ5suE8qr2sD8uWQ/pub)

**Comment by cuishuang** [comment]
"指向栈对象的指针不能存在于堆中；" 
应该是 "指向堆对象的指针不能存在于栈中"吧

---

*2020-10-15 UPDATES: 已修复*

*2020-10-27 UPDATES: 已回滚，原表述没有问题*

**Comment by bfwbbrj** [comment]
> @cuishuang
> "指向栈对象的指针不能存在于堆中；" 
> 应该是 "指向堆对象的指针不能存在于栈中"吧
> 
> ---
> 
> *2020-10-15 UPDATES: 已修复*

这个不应该是 “指向栈对象的指针不能存在于堆中“ 吗，之前的应该是对的吧

**Comment by draveness** [comment]
> > @cuishuang
> > "指向栈对象的指针不能存在于堆中；"
> > 应该是 "指向堆对象的指针不能存在于栈中"吧
> 
> 这个不应该是 “指向栈对象的指针不能存在于堆中“ 吗，之前的应该是对的吧

确实是，原来是对的

**Comment by mhh12121** [comment]
> @fengwutianya
> > > adjustpointer时怎么知道 这是栈上内容是指针而不是普通数据？
> > 
> > FYI
> > 
> > > Each word has 2 bits associated with it to tell you if it was a scalar or a pointer inside that word
> > > --- https://blog.golang.org/ismmkeynote
> 
> I found out that there is a metadata for every function which is stored in funcinfo

如果你是说adjustPointer而不是adjustPointers, 其第二个传入参数类型就是unsafe.Pointer，无论是sudogs.elem或者是sched.ctxt都是这个类型啊，意思就是这必是unsafe.Pointer? 而不用知道这个是不是普通数据或者指针？

**Comment by draveness** [comment]
>@mhh12121
> > @fengwutianya
> > > > adjustpointer时怎么知道 这是栈上内容是指针而不是普通数据？
> > > 
> > > 
> > > FYI
> > > > Each word has 2 bits associated with it to tell you if it was a scalar or a pointer inside that word
> > > > --- https://blog.golang.org/ismmkeynote
> > 
> > 
> > I found out that there is a metadata for every function which is stored in funcinfo
> 
> 如果你是说adjustPointer而不是adjustPointers, 其第二个传入参数类型就是unsafe.Pointer，无论是sudogs.elem或者是sched.ctxt都是这个类型啊，意思就是这必是unsafe.Pointer? 而不用知道这个是不是普通数据或者指针？

可以展开说一下你的问题么，感觉不是特别清晰

**Comment by mhh12121** [comment]
> > @mhh12121
> > > @fengwutianya
> > > > > adjustpointer时怎么知道 这是栈上内容是指针而不是普通数据？
> > > > 
> > > > 
> > > > FYI
> > > > > Each word has 2 bits associated with it to tell you if it was a scalar or a pointer inside that word
> > > > > --- https://blog.golang.org/ismmkeynote
> > > 
> > > 
> > > I found out that there is a metadata for every function which is stored in funcinfo
> > 
> > 
> > 如果你是说adjustPointer而不是adjustPointers, 其第二个传入参数类型就是unsafe.Pointer，无论是sudogs.elem或者是sched.ctxt都是这个类型啊，意思就是这必是unsafe.Pointer? 而不用知道这个是不是普通数据或者指针？
> 
> 可以展开说一下你的问题么，感觉不是特别清晰

实在不好意思，我其实同这位仁兄有同样的问题，如何得知这个是指针还是普通数据呢；在stack的copy中，我没有找到可以说明"判断这个是指针还是普通数据"的判断，我看见`adjustpointer(adjinfo *adjustinfo, vpp unsafe.Pointer)`以及其传入的参数，仿佛就assume了传入的一定是指针类型；
然后针对`funcinfo`的回答，我不明白，好像`adjustpointer`这个方法没有用到这个结构体？ 求解惑一下，十分感谢！

**Comment by draveness** [comment]
> > > @mhh12121
> > > > @fengwutianya
> > > > > > adjustpointer时怎么知道 这是栈上内容是指针而不是普通数据？
> > > > > 
> > > > > 
> > > > > FYI
> > > > > > Each word has 2 bits associated with it to tell you if it was a scalar or a pointer inside that word
> > > > > > --- https://blog.golang.org/ismmkeynote
> > > > 
> > > > 
> > > > I found out that there is a metadata for every function which is stored in funcinfo
> > > 
> > > 
> > > 如果你是说adjustPointer而不是adjustPointers, 其第二个传入参数类型就是unsafe.Pointer，无论是sudogs.elem或者是sched.ctxt都是这个类型啊，意思就是这必是unsafe.Pointer? 而不用知道这个是不是普通数据或者指针？
> > 
> > 
> > 可以展开说一下你的问题么，感觉不是特别清晰
> 

是有两个问题么

> 实在不好意思，我其实同这位仁兄有同样的问题，如何得知这个是指针还是普通数据呢；
> 在stack的copy中，我没有找到可以说明"判断这个是指针还是普通数据"的判断，我看见`adjustpointer(adjinfo *adjustinfo, vpp unsafe.Pointer)`以及其传入的参数，仿佛就assume了传入的一定是指针类型；

这个问题上面不是回复过了么？

> 然后针对`funcinfo`的回答，我不明白，好像`adjustpointer`这个方法没有用到这个结构体？ 求解惑一下，十分感谢！

针对 `funcinfo` 的回答，好像是 @fengwutianya 自问自答的，我也不清楚这里是啥意思

**Comment by ningmonguo** [comment]
请问一下m的g0会扩容吗？这一块是怎么处理的呢？

**Comment by draveness** [comment]
> 请问一下m的g0会扩容吗？这一块是怎么处理的呢？

如果你说的是 g0 的栈，g0 的栈是不会调 morestack 的，可以看这个 

https://github.com/golang/go/blob/532c199ee56cdbc2cfd12da1c1cfb3359b122c7c/src/runtime/asm_arm.s#L379

**Comment by ningmonguo** [comment]
> @draveness
> > 请问一下m的g0会扩容吗？这一块是怎么处理的呢？
> 
> 如果你说的是 g0 的栈，g0 的栈是不会调 morestack 的，可以看这个 
> 
> https://github.com/golang/go/blob/532c199ee56cdbc2cfd12da1c1cfb3359b122c7c/src/runtime/asm_arm.s#L379

感谢大神(/崇拜脸)，也就是g0栈是保证了 8k 够用的吗？

**Comment by MMMzq** [comment]
你好!
我有一些关于逃逸分析的的问题?
```go
func testSlice()[]uintptr{
	var r []uintptr
	
	for i := 0; i < 10; i++ {
		tc := &[10]byte{}
		r = append(r, uintptr(unsafe.Pointer(tc)))
	}

	return r
}
func main() {
	slice := testSlice()
	fmt.Printf("%d\n",slice[0])
	fmt.Printf("%d\n",slice[1])
	fmt.Printf("%d\n",slice[2])
}
```
结果:
```
 go run -gcflags '-m -l' .\t.go

.\t.go:20:9: &[10]byte literal does not escape
.\t.go:28:12: ... argument does not escape
.\t.go:28:25: slice[0] escapes to heap
.\t.go:29:12: ... argument does not escape
.\t.go:29:25: slice[1] escapes to heap
.\t.go:30:12: ... argument does not escape
.\t.go:30:25: slice[2] escapes to heap
824634236670
824634236670
824634236670
```

1.  uintptr和unsafe.Pointer是会影响逃逸分析的吗?
2.  为什么slice[0],slice[1],slice[2]他们的值相等?

对于第二点: 我自己的想法是,因为`tc`并没有逃逸,等下一轮循环的时候,栈顶指针又回到原来的位置,不知道对不对

**Comment by Imfan** [comment]
大神
设计原理 第一句话
>栈区的内存一般由编译器自动进行分配和释放

栈区的内存不是由 系统自动分配和释放的吗
或者说
栈区的内存一般由编译器通过指令自动进行分配和释放 。

**Comment by draveness** [comment]
> 大神
> 设计原理 第一句话
> 
> > 栈区的内存一般由编译器自动进行分配和释放
> 
> 栈区的内存不是由 系统自动分配和释放的吗

这里的系统分配非常宽泛

> 或者说
> 栈区的内存一般由编译器通过指令自动进行分配和释放 。

这种说法更精确一些，不过跟编译器管理差不多

**Comment by draveness** [comment]
> 你好!
> 我有一些关于逃逸分析的的问题?
> 
> ```go
> func testSlice()[]uintptr{
> 	var r []uintptr
> 	
> 	for i := 0; i < 10; i++ {
> 		tc := &[10]byte{}
> 		r = append(r, uintptr(unsafe.Pointer(tc)))
> 	}
> 
> 	return r
> }
> func main() {
> 	slice := testSlice()
> 	fmt.Printf("%d\n",slice[0])
> 	fmt.Printf("%d\n",slice[1])
> 	fmt.Printf("%d\n",slice[2])
> }
> ```
> 
> 结果:
> 
> ```
>  go run -gcflags '-m -l' .\t.go
> 
> .\t.go:20:9: &[10]byte literal does not escape
> .\t.go:28:12: ... argument does not escape
> .\t.go:28:25: slice[0] escapes to heap
> .\t.go:29:12: ... argument does not escape
> .\t.go:29:25: slice[1] escapes to heap
> .\t.go:30:12: ... argument does not escape
> .\t.go:30:25: slice[2] escapes to heap
> 824634236670
> 824634236670
> 824634236670
> ```
> 
> 1. uintptr和unsafe.Pointer是会影响逃逸分析的吗?

uintptr 主要会影响垃圾回收

> 2. 为什么slice[0],slice[1],slice[2]他们的值相等?
> 
> 对于第二点: 我自己的想法是,因为`tc`并没有逃逸,等下一轮循环的时候,栈顶指针又回到原来的位置,不知道对不对

你的想法应该没问题，可以编译成汇编看看具体实现

**Comment by Northward712** [comment]
大佬 今天跟同事讨论这个讨论到go的goroutine算是个有栈协程而类似lua,js这种则是无栈协程，那么为什么go没有选择无栈协程呢？这样不是避免了栈内存分配和移动等问题吗

**Comment by waittttting** [comment]
> 所以它在用户台实现 Goroutine 作为执行上下文。

用户态写成了用户台

---

*2021-01-30 UPDATES: 已修复*

**Comment by draveness** [comment]
> @E-c-h-o-Go 大佬 今天跟同事讨论这个讨论到go的goroutine算是个有栈协程而类似lua,js这种则是无栈协程，那么为什么go没有选择无栈协程呢？这样不是避免了栈内存分配和移动等问题吗

Go 语言的 Goroutine 不能和 Javascript 对比，它与操作系统的 Thread 其实更相似，没有调用栈的话，函数调用是很难挂起的，如果运行时突然要执行调度或者 GC 是需要停止当前的 Thread 或者 Goroutine

### 推荐阅读

+ https://www.chiark.greenend.org.uk/~sgtatham/coroutines.html

**Comment by schecterdamien** [comment]
大佬，栈缩容的过程是在gc里全部完成的吗，还是gc只是标记下需要缩容，然后在函数调用前的检查（runtime.newstack）里真正的执行缩容操作

**Comment by bugangongwei** [comment]
> @MMMzq
> 你好!
> 我有一些关于逃逸分析的的问题?
> ```go
> func testSlice()[]uintptr{
> 	var r []uintptr
> 	
> 	for i := 0; i < 10; i++ {
> 		tc := &[10]byte{}
> 		r = append(r, uintptr(unsafe.Pointer(tc)))
> 	}
> 
> 	return r
> }
> func main() {
> 	slice := testSlice()
> 	fmt.Printf("%d\n",slice[0])
> 	fmt.Printf("%d\n",slice[1])
> 	fmt.Printf("%d\n",slice[2])
> }
> ```
> 结果:
> ```
>  go run -gcflags '-m -l' .\t.go
> 
> .\t.go:20:9: &[10]byte literal does not escape
> .\t.go:28:12: ... argument does not escape
> .\t.go:28:25: slice[0] escapes to heap
> .\t.go:29:12: ... argument does not escape
> .\t.go:29:25: slice[1] escapes to heap
> .\t.go:30:12: ... argument does not escape
> .\t.go:30:25: slice[2] escapes to heap
> 824634236670
> 824634236670
> 824634236670
> ```
> 
> 1.  uintptr和unsafe.Pointer是会影响逃逸分析的吗?
> 2.  为什么slice[0],slice[1],slice[2]他们的值相等?
> 
> 对于第二点: 我自己的想法是,因为`tc`并没有逃逸,等下一轮循环的时候,栈顶指针又回到原来的位置,不知道对不对
> 

您好, 关于第一个问题, 我恰好有看到一篇文章 https://zhuanlan.zhihu.com/p/91559562; 
其实 ``slice[0]``, ``slice[1]``, ``slice[2]`` 的逃逸是因为 fmt.Printf, 查看源码
````go
func Printf(format string, a ...interface{}) (n int, err error) {
  // a 就是要打印的值, 比如 slice[0]
  return Fprintf(os.Stdout, format, a...)
}

func Fprintf(w io.Writer, format string, a ...interface{}) (n int, err error) {
  // p 因为是函数 newPrinter() 返回的指针, 函数 newPrinter() 栈已经释放了, 但是 p 还需要被使用, 所以 p 肯定是逃逸了的
  // 情况 1: 通过构造函数返回的指针一定是逃逸的
  p := newPrinter()
  p.doPrintf(format, a)
  n, err = w.Write(p.buf)
  p.free()
  return
}

func (p *pp) doPrintf(format string, a []interface{}) {
  ....
  p.printArg(a[argNum], verb)
  argNum++
  ....
}


func (p *pp) printArg(arg interface{}, verb rune) {
  // p 通过 arg 引用了 arg(a[argNum])
  // 情况2: 被已经逃逸的指针指向的指针, 也是逃逸的
  p.arg = arg
  ....
}
````
以上, 就是第一个问题, ``slice[0]``, ``slice[1]``, ``slice[2]`` 逃逸的原因;

**Comment by ThreeBearsDan** [comment]
> @mhh12121
> > > @mhh12121
> > > > @fengwutianya
> > > > > > adjustpointer时怎么知道 这是栈上内容是指针而不是普通数据？
> > > > > 
> > > > > 
> > > > > FYI
> > > > > > Each word has 2 bits associated with it to tell you if it was a scalar or a pointer inside that word
> > > > > > --- https://blog.golang.org/ismmkeynote
> > > > 
> > > > 
> > > > I found out that there is a metadata for every function which is stored in funcinfo
> > > 
> > > 
> > > 如果你是说adjustPointer而不是adjustPointers, 其第二个传入参数类型就是unsafe.Pointer，无论是sudogs.elem或者是sched.ctxt都是这个类型啊，意思就是这必是unsafe.Pointer? 而不用知道这个是不是普通数据或者指针？
> > 
> > 可以展开说一下你的问题么，感觉不是特别清晰
> 
> 实在不好意思，我其实同这位仁兄有同样的问题，如何得知这个是指针还是普通数据呢；在stack的copy中，我没有找到可以说明"判断这个是指针还是普通数据"的判断，我看见`adjustpointer(adjinfo *adjustinfo, vpp unsafe.Pointer)`以及其传入的参数，仿佛就assume了传入的一定是指针类型；
> 然后针对`funcinfo`的回答，我不明白，好像`adjustpointer`这个方法没有用到这个结构体？ 求解惑一下，十分感谢！

[答案在这里](https://www.altoros.com/blog/golang-internals-part-4-object-files-and-function-metadata/) 最后一部分

**Comment by xxxxzgxxxx** [comment]
大佬，连续栈的扩容操作步骤2 ： 将旧栈中的所有内容复制到新栈中但是实体书上的内容是「将旧堆中的所有内容复制到新堆中」

**Reply by draveness** [reply]
这个实体书估计写错了，我跟编辑说一下

**Comment by HEMOURENNN** [comment]
“我们可以认为 Go 语言的栈内存都是分配在堆上的”
大佬，这个我没懂是啥意思

**Reply by draveness** [reply]
因为栈相关的结构都存储在 runtime.mspan 中

**Reply by HEMOURENNN** [reply]
那也就是可以说是，栈其实是堆中的一块连续的内存空间。只是对于两者的处理方式不一样对哇？

**Reply by draveness** [reply]
你看一下 stackinit 函数，了解下栈是怎么初始化的，大概就明白了

**Reply by HEMOURENNN** [reply]
好的，谢谢大佬

**Reply by opfsun** [reply]
栈内存函数返回就回收了，如果栈是堆的话，就需要gc回收了。

**Comment by chenfengjin** [comment]
关于“寄存器在物理机上非常有限，然而栈区的操作会使用到两个以上的寄存器，这足以说明栈内存在应用程序的重要性”的一点补充说明
从指令集设计的角度来看，指令集可以分为基于栈的指令集结构和寄存器的指令结构。在某些(虚拟)指令集结构中，只有栈而没有寄存器(Lua虚拟机，WebAssembly 虚拟机)。
“栈区的操作会使用到两个以上的寄存器”只是针对 x86 体系结构而言，事实上，在X86体系结构上内存访问都基本会涉及到两个甚至多个寄存器，如普通数据段访问依赖段选择子和段内偏移

**Comment by yangningkit** [comment]
关于“需要注意的是，Go 语言的编译器不会为所有的函数插入 runtime.morestack，它只会在必要时插入指令以减少运行时的额外开销，编译指令 nosplit 可以跳过栈溢出的检查，虽然这能降低一些开销，不过固定大小的栈也存在溢出的风险。”有点疑惑
在cmd/internal/obj/x86.preprocess函数中仅调用栈分裂检查cmd/internal/obj/x86.stacksplit函数前，条件仅排除了添加了nosplit注释指的函数，那上文中“不会为所有的函数插入”这里的条件就只有添加了nosplit指令才会跳过栈溢出的检查吗

**Comment by chenwei67** [comment]
大佬你好。我有个疑问是go的栈和操作系统如linux的栈是怎样的关系？操作系统的栈是由操作系统管理，Go的运行时不能操作吧？那Go的栈是全在堆上吗？操作系统的栈在Go中有作用吗？作用是什么？
不好意思问题有点多，希望大佬看到了能赐教~

**Comment by itcuijian** [comment]
> `runtime.stackpoolalloc` 会在全局的栈缓存池 `runtime.stackpool` 中获取新的内存，如果栈缓存池中不包含剩余的内存，运行时会从堆上申请一片内存空间；如果线程缓存中包含足够的空间，我们可以从线程本地的缓存中获取内存，一旦发现空间不足就会调用 `runtime.stackcacherefill` 从堆上获取新的内存。
这个说法不对，并不会在线程缓存中获取内存，`runtime.stackcacherefill` 函数内部实际上也是调用 `runtime.stackpoolalloc` 为全局变量 `runtime.stackpool` 申请一块新空间。
